app/com/arpnetworking/kairos/service/KairosDbServiceImpl.java

```diff
@@ -299,6 +297,44 @@ private static MetricsQuery useAvailableRollups(
                 }).collect(ImmutableList.toImmutableList())));
     }
 
+    /**
+     * Get the coarsest rollup time period that could be used to rewrite a query without affecting its results.
+     *
+     * @param metric The metric-query that we might want to rewrite to make use of rollups.
+     * @return The largest {@link SamplingUnit} that the query is insensitive to aggregation over
+     *     (or {@code empty}, if the query is sensitive to all pre-aggregation).
+     */
+    /* package private */ static Optional<SamplingUnit> getMaxUsableRollupUnit(final Metric metric) {
+        /* By querying a rollup (say, the hourly rollup), we're effectively putting a sampling-aligned `merge(1h)`
+         *   in front of the metric's aggregators.
+         * For this to not affect the results, the first aggregator must be sampling-aligned, and must aggregate over
+         *   some integer number of hours.
+         * (Empirically, non-sampling aggregators (e.g. `div`, `filter`) commute with `merge`, so we can ignore them.)
```

Comment:
nit. For this comment and all of the below, `sampling aggregator` should be replaced with `range aggregator` to be consistent with the KairosDB terminology

i.e. only range aggregators have a `sampling` argument https://kairosdb.github.io/docs/build/html/restapi/Aggregators.html#range-aggregator

Reply:
Ah, yes, thank you! I did feel fuzzy on that terminology.